Add blog for kep-3721 (#51529)
kubernetes/website: content/en/blog/_posts/2025-09-01-introducing-env-files/index.md

---

> layout: blog

Reviewer: The blog post is well-written giving out the feature aspects and conversational making it easily readable. Thanks for your contribution!

title: "Kubernetes v1.34: Introducing Environment variable files"
date: 2025-0X-XX
draft: true
slug: kubernetes-v1-34-env-files
author: >
  HirazawaUi
---

## EnvFiles: A New Approach to Kubernetes Environment Variables
Kubernetes typically uses `ConfigMaps` and `Secrets` to set environment variables, 

> which introduces additional API calls and complexity. 

Author: I've added some reasoning, though it might not align with what you expected. Could you take a look?

Alternatively, you might be using a vendor-supplied container 
that requires environment variables (such as a license key or a one-time token),
 but you don’t want to hard-code them or mount volumes just to get the job done.

At this point, you can use the env files feature, 
which is managed by the `EnvFiles` feature gate. 
Currently in the Alpha stage,
this functionality allows you to load environment variables directly from a file in an emptyDir volume
without actually mounting that file into the container.
It’s a simple yet elegant solution to some surprisingly common problems.

## What’s EnvFiles All About?
At its core, `EnvFiles` allows you to point your container to a file,
one generated by an initContainer,
and have Kubernetes parse that file to set your environment variables.
The file lives in an emptyDir volume (a temporary storage space that lasts as long as the pod does),
Your main container doesn’t need to mount the volume.

> The kubelet will reading the file and injecting the variables when the container starts.

Reviewer: ```suggestion
The kubelet will read the file and inject these variables when the container starts.
```

## How It Works
Here's a simple example:
```
apiVersion: v1
kind: Pod
spec:
  initContainers:
  - name: generate-config
    image: busybox

>     command: ['sh', '-c', 'echo "CONFIG_VAR=HELLO" > /config/config.env']

Reviewer: Optional> replacing CONFIG_VAR with a functional example like looking up MDS feels more suited here for an example? wdyt?

```suggestion
    command: ['sh', '-c', 'echo "TOKEN=$(curl -s http://169.254.169.254/metadata/token)" > /config/config.env']
```

    volumeMounts:
    - name: config-volume
      mountPath: /config
  containers:
  - name: app-container
    image: gcr.io/distroless/static
    env:
    - name: CONFIG_VAR
      valueFrom:
        fileKeyRef:
          path: config.env
          volumeName: config-volume
          key: CONFIG_VAR
  volumes:
  - name: config-volume
    emptyDir: {}
```

Using EnvFiles is a breeze. 
You define your environment variables in the pod spec using the `fileKeyRef` field,
which tells Kubernetes where to find the file and which key to pull.
The file itself follows standard .env syntax (think KEY=VALUE),
and it’s stored in an emptyDir volume. 
The initContainer mounts the volume to write the file, 
but the main container doesn’t need to—it just gets the variables handed to it at startup.

## A Word on Security
One quick heads-up: while `EnvFiles` can handle sensitive data like keys or tokens, 
it’s not a drop-in replacement for Secrets. 

> The emptyDir volume doesn’t offer the same protections, 

Reviewer: This feels important. Consider adding a sentence to explain why emptyDir is not a replacement for Secrets (e.g., they aren't encrypted and are not subject to the same access controls).

---
layout: blog

> title: "Kubernetes v1.34: Introducing Environment variable files"

Reviewer 2: Consider changing the title to describe _what you can now do_, instead of the technological change that makes it possible.

For example:
- Kubernetes v1.34 Lets You Load Environment Variables From The Filesystem
- Kubernetes v1.34: Use An Init Container To Define App Environment Variables
- etc

date: 2025-0X-XX
draft: true
slug: kubernetes-v1-34-env-files
author: >
  HirazawaUi
---

> ## EnvFiles: A New Approach to Kubernetes Environment Variables

Reviewer 2: Actually, we avoid starting articles with a heading.

> ## EnvFiles: A New Approach to Kubernetes Environment Variables

Author: I've removed the title :) – it seems better to introduce it with a piece of text instead.

> Kubernetes typically uses `ConfigMaps` and `Secrets` to set environment variables, 

Reviewer 3: ```suggestion
Kubernetes typically uses ConfigMaps and Secrets to set environment variables, 
```

The style guide suggest to use upper camel case for API objects and not code format: https://kubernetes.io/docs/contribute/style/style-guide/#use-upper-camel-case-for-api-objects

and their configurations, while ensuring orderly 
updates for both the configurations and the workload Pods.

Alternatively, you might be using a vendor-supplied container 
that requires environment variables (such as a license key or a one-time token),
but you don’t want to hard-code them or mount volumes just to get the job done.

At this point, you can use the env files feature, 
which is managed by the `EnvFiles` feature gate. 
Currently in the Alpha stage,
this functionality allows you to load environment variables directly from a file in an emptyDir volume
without actually mounting that file into the container.
It’s a simple yet elegant solution to some surprisingly common problems.

## What’s EnvFiles All About?
At its core, `EnvFiles` allows you to point your container to a file,
one generated by an `initContainer`,
and have Kubernetes parse that file to set your environment variables.
The file lives in an `emptyDir` volume (a temporary storage space that lasts as long as the pod does),
Your main container doesn’t need to mount the volume.
The kubelet will read the file and inject these variables when the container starts.

## How It Works
Here's a simple example:

> ```

Reviewer 4: ```suggestion
```yaml
```

apiVersion: v1
kind: Pod
spec:
  initContainers:
  - name: generate-config
    image: busybox
    command: ['sh', '-c', 'echo "CONFIG_VAR=HELLO" > /config/config.env']
    volumeMounts:
    - name: config-volume
      mountPath: /config
  containers:
  - name: app-container
    image: gcr.io/distroless/static
    env:
    - name: CONFIG_VAR
      valueFrom:
        fileKeyRef:
          path: config.env
          volumeName: config-volume
          key: CONFIG_VAR
  volumes:
  - name: config-volume
    emptyDir: {}
```

Using EnvFiles is a breeze. 
You define your environment variables in the pod spec using the `fileKeyRef` field,
which tells Kubernetes where to find the file and which key to pull.

> The file itself follows standard .env syntax (think KEY=VALUE),

Reviewer 2: ```suggestion
The file itself follows the de-facto standard for `.env` syntax (think `KEY=VALUE`),
```

> and it’s stored in an emptyDir volume. 

Reviewer 2: ```suggestion
and (for this alpha stage at least) you must ensure that it is written into
an `emptyDir` volume. Other volume types aren't supported for this feature.
```

> The initContainer mounts the volume to write the file, 

Reviewer 2: ```suggestion
At least one init container must mount that `emptyDir` volume (to write the file),
```

but the main container doesn’t need to—it just gets the variables handed to it at startup.

## A Word on Security

> One quick heads-up: while `EnvFiles` can handle sensitive data like keys or tokens, 

Reviewer 2: Avoid using the feature gate string in the announcement, apart from the specific place where we give advice about enabling it.

it’s not a drop-in replacement for Secrets. 
The `emptyDir` volumes do not provide the same [security protections as Secrets](https://kubernetes.io/docs/concepts/configuration/secret/#information-security-for-secrets).
so if security is a top concern, stick with Secrets where possible.

> We’ll make sure the docs call this out clearly.

Reviewer 2: This is really bad advice. Using this new mechanism may actually provide better information security that Secret.

Only include this statement if you can get a SIG Security person, ideally a tech lead for SIG Security, to endorse that inclusion.

Author: Can I avoid involving the SIG Security maintainers by removing the description of the secret? They aren’t familiar with this feature and might unnecessarily spend extra time understanding it :)

---
layout: blog

> title: "Kubernetes 1.34: Introducing Environment variable files"

Reviewer 4: ```suggestion
title: "Kubernetes v1.34: Introducing Environment variable files"
```

> date: 2025-09-01

Reviewer 4: ```suggestion
date: 2025-0X-XX
draft: true
```

> slug: kubernetes-1-34-env-files

Reviewer 4: ```suggestion
slug: kubernetes-v1-34-env-files
```
